packages/ragbits-core/src/ragbits/core/utils/dict_transformations.py

```diff
@@ -0,0 +1,145 @@
+from typing import Any
+
+
+def flatten_dict(
+    input_dict: dict[str, Any], parent_key: str = "", sep: str = "."
+) -> dict[str, str | int | float | bool]:
+    """
+    Recursively flatten a nested dictionary and lists, converting non-primitive types to strings.
+
+    Args:
+        input_dict: The dictionary to flatten
+        parent_key: The parent key for nested keys
+        sep: The separator between nested keys
+
+    Returns:
+        A flattened dictionary with primitive types and stringified complex types
+
+    Raises:
+        ValueError: If the dictionary cannot be safely flattened due to the presence of the separator in the dict key.
+    """
+    items: dict[str, str | int | float | bool] = {}
+    for k, v in input_dict.items():
+        if sep in k:
+            raise ValueError(f"Separator '{sep}' found in key '{parent_key}' Cannot flatten dictionary safely.")
+
+        if k.endswith("]"):
+            raise ValueError(f"Key '{k}' cannot end with ']' Cannot flatten dictionary safely.")
+
+        new_key = f"{parent_key}{sep}{k}" if parent_key else k
+
+        if isinstance(v, dict):
+            items = {**items, **flatten_dict(v, new_key, sep=sep)}
+        elif isinstance(v, list):
+            for i, item in enumerate(v):
+                list_key = f"{new_key}[{i}]"
+                if isinstance(item, dict):
+                    items = {**items, **flatten_dict(item, list_key, sep=sep)}
+                else:
+                    items[list_key] = item if isinstance(item, str | int | float | bool) else str(item)
+        else:
+            items[new_key] = v if isinstance(v, str | int | float | bool) else str(v)
+
+    return items
+
+
+def unflatten_dict(input_dict: dict[Any, Any]) -> dict[Any, Any] | list:
```

Comment:
`dict[Any, Any]`? Why not `dict[str, Any]` or `dict[str, str | int | float | bool | dict]`?

Reply:
fixed